test/acceptance/examples/main_test.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package examples
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	terratestLogger "github.com/gruntwork-io/terratest/modules/logger"
+	"github.com/gruntwork-io/terratest/modules/random"
+	"github.com/gruntwork-io/terratest/modules/terraform"
+	"github.com/hashicorp/consul/sdk/testutil/retry"
+	"github.com/hashicorp/terraform-aws-consul-ecs/test/acceptance/examples/scenarios"
+	clusterpeering "github.com/hashicorp/terraform-aws-consul-ecs/test/acceptance/examples/scenarios/cluster-peering"
+	"github.com/hashicorp/terraform-aws-consul-ecs/test/acceptance/examples/scenarios/ec2"
+	"github.com/hashicorp/terraform-aws-consul-ecs/test/acceptance/examples/scenarios/fargate"
+	"github.com/hashicorp/terraform-aws-consul-ecs/test/acceptance/examples/scenarios/hcp"
+	localityawarerouting "github.com/hashicorp/terraform-aws-consul-ecs/test/acceptance/examples/scenarios/locality-aware-routing"
+	sameness "github.com/hashicorp/terraform-aws-consul-ecs/test/acceptance/examples/scenarios/service-sameness"
+	"github.com/hashicorp/terraform-aws-consul-ecs/test/acceptance/examples/scenarios/wan-federation"
+	"github.com/hashicorp/terraform-aws-consul-ecs/test/acceptance/framework/logger"
+	"github.com/stretchr/testify/require"
+)
+
+type ScenarioFunc func(string) scenarios.Scenario
+
+var scenarioFuncs = map[string]ScenarioFunc{
+	"FARGATE":                fargate.New,
+	"EC2":                    ec2.New,
+	"HCP":                    hcp.New,
+	"CLUSTER_PEERING":        clusterpeering.New,
+	"WAN_FEDERATION":         wan.New,
+	"LOCALITY_AWARE_ROUTING": localityawarerouting.New,
+	"SERVICE_SAMENESS":       sameness.New,
+}
```

Comment:
Instead of hard coding the strings here will it be better to just keep array of functions. and create this scenarioFuncs dynamically like this. I have created a patch file please check.
[Uploading changes.patch…]()

Reply:
```
diff --git a/test/acceptance/examples/main_test.go b/test/acceptance/examples/main_test.go
index d6865b0..a684c98 100644
--- a/test/acceptance/examples/main_test.go
+++ b/test/acceptance/examples/main_test.go
@@ -6,12 +6,10 @@ package examples
 import (
 	"fmt"
 	"os"
-	"strings"
 	"testing"
 	"time"

 	terratestLogger "github.com/gruntwork-io/terratest/modules/logger"
-	"github.com/gruntwork-io/terratest/modules/random"
 	"github.com/gruntwork-io/terratest/modules/terraform"
 	"github.com/hashicorp/consul/sdk/testutil/retry"
 	"github.com/hashicorp/terraform-aws-consul-ecs/test/acceptance/examples/scenarios"
@@ -26,26 +24,36 @@ import (
 	"github.com/stretchr/testify/require"
 )

-type ScenarioFunc func(string) scenarios.Scenario
+type ScenarioFunc func() (scenarios.Scenario, string)

-var scenarioFuncs = map[string]ScenarioFunc{
-	"FARGATE":                fargate.New,
-	"EC2":                    ec2.New,
-	"HCP":                    hcp.New,
-	"CLUSTER_PEERING":        clusterpeering.New,
-	"WAN_FEDERATION":         wan.New,
-	"LOCALITY_AWARE_ROUTING": localityawarerouting.New,
-	"SERVICE_SAMENESS":       sameness.New,
+var scenarioFuncs = []ScenarioFunc{
+	fargate.New,
+	ec2.New,
+	hcp.New,
+	clusterpeering.New,
+	wan.New,
+	localityawarerouting.New,
+	sameness.New,
+}
+
+var scenarioNameFuncMap map[string]scenarios.Scenario
+
+func setupScenarios() {
+	for _, scenarioFunc := range scenarioFuncs {
+		scenarioObj, name := scenarioFunc()
+		scenarioNameFuncMap[name] = scenarioObj
+	}
 }

 // TestRunScenario accepts a single scenario name as an environment
 // variable and executes tests for the same. We want to run each
 // scenario as a separate GitHub Action job.
 func TestRunScenario(t *testing.T) {
+	setupScenarios()
 	scenarioName := os.Getenv("TEST_SCENARIO")
 	require.NotEmpty(t, scenarioName)

-	scenario := getScenario(scenarioName)
+	scenario := scenarioNameFuncMap[scenarioName]
 	require.NotNil(t, scenario, "unexpected scenario name")

 	initOptions := terraform.WithDefaultRetryableErrors(t, &terraform.Options{
@@ -84,16 +92,3 @@ func TestRunScenario(t *testing.T) {
 	scenario.Validate(t, outputs)
 	logger.Log(t, "validation successful!!")
 }
-
-func getScenario(name string) scenarios.Scenario {
-	// This name will be passed as the name input to the terraform apply call.
-	// This name can be modified inside the individual scenarios too according
-	// to the scenario's needs.
-	terraformResourcesName := fmt.Sprintf("ecs-%s", strings.ToLower(random.UniqueId()))
-
-	if scenario, ok := scenarioFuncs[name]; ok {
-		return scenario(terraformResourcesName)
-	}
-
-	return nil
-}
diff --git a/test/acceptance/examples/scenarios/cluster-peering/main.go b/test/acceptance/examples/scenarios/cluster-peering/main.go
index c1f190f..3598322 100644
--- a/test/acceptance/examples/scenarios/cluster-peering/main.go
+++ b/test/acceptance/examples/scenarios/cluster-peering/main.go
@@ -14,16 +14,20 @@ import (
 	"github.com/stretchr/testify/require"
 )

-type clusterPeering struct {
-	name string
-}
+type clusterPeering struct{}

-func New(name string) scenarios.Scenario {
+func New() (scenarios.Scenario, string) {
 	// We go ahead with a simple name due to AWS restrictions
 	// on character length of resource names for certain resources.
-	return &clusterPeering{
-		name: "ecs",
-	}
+	return &clusterPeering{}, GetScenarioName()
+}
+
+func GetScenarioName() string {
+	return "CLUSTER_PEERING"
+}
+
+func (c *clusterPeering) GetTerraformResourcesName() string {
+	return "ecs"
 }

 func (c *clusterPeering) GetFolderName() string {
@@ -33,7 +37,7 @@ func (c *clusterPeering) GetFolderName() string {
 func (c *clusterPeering) GetTerraformVars() (map[string]interface{}, error) {
 	vars := map[string]interface{}{
 		"region": "us-east-2",
-		"name":   c.name,
+		"name":   c.GetTerraformVars(),
 	}

 	publicIP, err := common.GetPublicIP()
@@ -68,13 +72,13 @@ func (c *clusterPeering) Validate(t *testing.T, outputVars map[string]interface{
 	consulClientTwo, err := common.SetupConsulClient(t, dc2ConsulServerURL, common.WithToken(dc2ConsulServerToken))
 	require.NoError(t, err)

-	clientAppName := fmt.Sprintf("%s-dc1-example-client-app", c.name)
-	serverAppName := fmt.Sprintf("%s-dc2-example-server-app", c.name)
+	clientAppName := fmt.Sprintf("%s-dc1-example-client-app", c.GetTerraformResourcesName())
+	serverAppName := fmt.Sprintf("%s-dc2-example-server-app", c.GetTerraformResourcesName())

 	consulClientOne.EnsureServiceReadiness(clientAppName, nil)
 	consulClientTwo.EnsureServiceReadiness(serverAppName, nil)
-	consulClientOne.EnsureServiceReadiness(fmt.Sprintf("%s-dc1-mesh-gateway", c.name), nil)
-	consulClientTwo.EnsureServiceReadiness(fmt.Sprintf("%s-dc2-mesh-gateway", c.name), nil)
+	consulClientOne.EnsureServiceReadiness(fmt.Sprintf("%s-dc1-mesh-gateway", c.GetTerraformResourcesName()), nil)
+	consulClientTwo.EnsureServiceReadiness(fmt.Sprintf("%s-dc2-mesh-gateway", c.GetTerraformResourcesName()), nil)

 	// Perform assertions by hitting the client app's LB
 	logger.Log(t, "calling client app's load balancer to see if the server app in the peer cluster is reachable")
diff --git a/test/acceptance/examples/scenarios/ec2/main.go b/test/acceptance/examples/scenarios/ec2/main.go
index 7b4f7af..f0b05fa 100644
--- a/test/acceptance/examples/scenarios/ec2/main.go
+++ b/test/acceptance/examples/scenarios/ec2/main.go
@@ -5,6 +5,7 @@ package ec2

 import (
 	"fmt"
+	"github.com/gruntwork-io/terratest/modules/random"
 	"strings"
 	"testing"

@@ -14,24 +15,28 @@ import (
 	"github.com/stretchr/testify/require"
 )

-type ec2 struct {
-	name string
+type ec2 struct{}
+
+func New() (scenarios.Scenario, string) {
+	return &ec2{}, GetScenarioName()
 }

-func New(name string) scenarios.Scenario {
-	return &ec2{
-		name: name,
-	}
+func GetScenarioName() string {
+	return "EC2"
 }

 func (e *ec2) GetFolderName() string {
 	return "dev-server-ec2"
 }

+func (e *ec2) GetTerraformResourcesName() string {
+	return fmt.Sprintf("ecs-%s", strings.ToLower(random.UniqueId()))
+}
+
 func (e *ec2) GetTerraformVars() (map[string]interface{}, error) {
 	vars := map[string]interface{}{
 		"region": "us-east-2",
-		"name":   e.name,
+		"name":   e.GetTerraformResourcesName(),
 	}

 	publicIP, err := common.GetPublicIP()
@@ -59,8 +64,8 @@ func (e *ec2) Validate(t *testing.T, outputVars map[string]interface{}) {
 	consulClient, err := common.SetupConsulClient(t, consulServerLBAddr)
 	require.NoError(t, err)

-	clientAppName := fmt.Sprintf("%s-example-client-app", e.name)
-	serverAppName := fmt.Sprintf("%s-example-server-app", e.name)
+	clientAppName := fmt.Sprintf("%s-example-client-app", e.GetTerraformResourcesName())
+	serverAppName := fmt.Sprintf("%s-example-server-app", e.GetTerraformResourcesName())

 	consulClient.EnsureServiceReadiness(clientAppName, nil)
 	consulClient.EnsureServiceReadiness(serverAppName, nil)
diff --git a/test/acceptance/examples/scenarios/fargate/main.go b/test/acceptance/examples/scenarios/fargate/main.go
index b847ce4..4642127 100644
--- a/test/acceptance/examples/scenarios/fargate/main.go
+++ b/test/acceptance/examples/scenarios/fargate/main.go
@@ -5,6 +5,7 @@ package fargate

 import (
 	"fmt"
+	"github.com/gruntwork-io/terratest/modules/random"
 	"strings"
 	"testing"

@@ -14,14 +15,18 @@ import (
 	"github.com/stretchr/testify/require"
 )

-type fargate struct {
-	name string
+type fargate struct{}
+
+func New() (scenarios.Scenario, string) {
+	return &fargate{}, GetScenarioName()
 }

-func New(name string) scenarios.Scenario {
-	return &fargate{
-		name: name,
-	}
+func GetScenarioName() string {
+	return "FARGATE"
+}
+
+func (f *fargate) GetTerraformResourcesName() string {
+	return fmt.Sprintf("ecs-%s", strings.ToLower(random.UniqueId()))
 }

 func (f *fargate) GetFolderName() string {
@@ -31,7 +36,7 @@ func (f *fargate) GetFolderName() string {
 func (f *fargate) GetTerraformVars() (map[string]interface{}, error) {
 	vars := map[string]interface{}{
 		"region": "us-east-1",
-		"name":   f.name,
+		"name":   f.GetTerraformResourcesName(),
 	}

 	publicIP, err := common.GetPublicIP()
@@ -59,8 +64,8 @@ func (f *fargate) Validate(t *testing.T, outputVars map[string]interface{}) {
 	consulClient, err := common.SetupConsulClient(t, consulServerLBAddr)
 	require.NoError(t, err)

-	clientAppName := fmt.Sprintf("%s-example-client-app", f.name)
-	serverAppName := fmt.Sprintf("%s-example-server-app", f.name)
+	clientAppName := fmt.Sprintf("%s-example-client-app", f.GetTerraformResourcesName())
+	serverAppName := fmt.Sprintf("%s-example-server-app", f.GetTerraformResourcesName())

 	consulClient.EnsureServiceReadiness(clientAppName, nil)
 	consulClient.EnsureServiceReadiness(serverAppName, nil)
diff --git a/test/acceptance/examples/scenarios/hcp/main.go b/test/acceptance/examples/scenarios/hcp/main.go
index b15eea7..c5cee2d 100644
--- a/test/acceptance/examples/scenarios/hcp/main.go
+++ b/test/acceptance/examples/scenarios/hcp/main.go
@@ -28,8 +28,16 @@ type service struct {
 	namespace     string
 }

-func New(name string) scenarios.Scenario {
-	return &hcp{}
+func New() (scenarios.Scenario, string) {
+	return &hcp{}, GetScenarioName()
+}
+
+func GetScenarioName() string {
+	return "HCP"
+}
+
+func (h *hcp) GetTerraformResourcesName() string {
+	return ""
 }

 func (h *hcp) GetFolderName() string {
diff --git a/test/acceptance/examples/scenarios/locality-aware-routing/main.go b/test/acceptance/examples/scenarios/locality-aware-routing/main.go
index ee1e41a..9789304 100644
--- a/test/acceptance/examples/scenarios/locality-aware-routing/main.go
+++ b/test/acceptance/examples/scenarios/locality-aware-routing/main.go
@@ -18,24 +18,28 @@ import (
 	"github.com/stretchr/testify/require"
 )

-type localityAwareRouting struct {
-	name string
+type localityAwareRouting struct{}
+
+func New() (scenarios.Scenario, string) {
+	return &localityAwareRouting{}, GetScenarioName()
 }

-func New(name string) scenarios.Scenario {
-	return &localityAwareRouting{
-		name: "con",
-	}
+func GetScenarioName() string {
+	return "LOCALITY_AWARE_ROUTING"
 }

 func (l *localityAwareRouting) GetFolderName() string {
 	return "locality-aware-routing"
 }

+func (l *localityAwareRouting) GetTerraformResourcesName() string {
+	return "con"
+}
+
 func (l *localityAwareRouting) GetTerraformVars() (map[string]interface{}, error) {
 	vars := map[string]interface{}{
 		"region": "us-west-2",
-		"name":   l.name,
+		"name":   l.GetTerraformResourcesName(),
 	}

 	publicIP, err := common.GetPublicIP()
diff --git a/test/acceptance/examples/scenarios/scenario.go b/test/acceptance/examples/scenarios/scenario.go
index 10ba57c..292c8a9 100644
--- a/test/acceptance/examples/scenarios/scenario.go
+++ b/test/acceptance/examples/scenarios/scenario.go
@@ -18,4 +18,6 @@ type Scenario interface {
 	// Validations that needs to be performed on the deployment. This
 	// hook will only be called after a successful terraform apply.
 	Validate(t *testing.T, tfOutput map[string]interface{})
+
+	GetTerraformResourcesName() string
 }
diff --git a/test/acceptance/examples/scenarios/service-sameness/main.go b/test/acceptance/examples/scenarios/service-sameness/main.go
index 25fea93..3509553 100644
--- a/test/acceptance/examples/scenarios/service-sameness/main.go
+++ b/test/acceptance/examples/scenarios/service-sameness/main.go
@@ -18,28 +18,32 @@ import (
 	"github.com/stretchr/testify/require"
 )

-type sameness struct {
-	name string
-}
+type sameness struct{}

 const (
 	awsRegion = "us-west-1"
 )

-func New(name string) scenarios.Scenario {
-	return &sameness{
-		name: "same",
-	}
+func New() (scenarios.Scenario, string) {
+	return &sameness{}, GetScenarioName()
+}
+
+func GetScenarioName() string {
+	return "SERVICE_SAMENESS"
 }

 func (s *sameness) GetFolderName() string {
 	return "service-sameness"
 }

+func (s *sameness) GetTerraformResourcesName() string {
+	return "same"
+}
+
 func (s *sameness) GetTerraformVars() (map[string]interface{}, error) {
 	vars := map[string]interface{}{
 		"region": awsRegion,
-		"name":   s.name,
+		"name":   s.GetTerraformResourcesName(),
 	}

 	publicIP, err := common.GetPublicIP()
@@ -90,9 +94,9 @@ func (s *sameness) Validate(t *testing.T, outputVars map[string]interface{}) {
 	ensureAppsReadiness(t, consulClientTwo, dc2DefaultPartitionApps)

 	// Ensure that the gateways are also ready
-	consulClientOne.EnsureServiceReadiness(fmt.Sprintf("%s-dc1-default-mesh-gateway", s.name), nil)
-	consulClientOne.EnsureServiceReadiness(fmt.Sprintf("%s-dc1-%s-mesh-gateway", s.name, dc1Part1PartitionApps.partition), &api.QueryOptions{Partition: dc1Part1PartitionApps.partition})
-	consulClientTwo.EnsureServiceReadiness(fmt.Sprintf("%s-dc2-mesh-gateway", s.name), nil)
+	consulClientOne.EnsureServiceReadiness(fmt.Sprintf("%s-dc1-default-mesh-gateway", s.GetTerraformResourcesName()), nil)
+	consulClientOne.EnsureServiceReadiness(fmt.Sprintf("%s-dc1-%s-mesh-gateway", s.GetTerraformResourcesName(), dc1Part1PartitionApps.partition), &api.QueryOptions{Partition: dc1Part1PartitionApps.partition})
+	consulClientTwo.EnsureServiceReadiness(fmt.Sprintf("%s-dc2-mesh-gateway", s.GetTerraformResourcesName()), nil)

 	// Begin actual validation
 	clusterAppsList := []*apps{
diff --git a/test/acceptance/examples/scenarios/wan-federation/main.go b/test/acceptance/examples/scenarios/wan-federation/main.go
index 9c06673..4ef262d 100644
--- a/test/acceptance/examples/scenarios/wan-federation/main.go
+++ b/test/acceptance/examples/scenarios/wan-federation/main.go
@@ -14,24 +14,28 @@ import (
 	"github.com/stretchr/testify/require"
 )

-type wan struct {
-	name string
+type wan struct{}
+
+func New() (scenarios.Scenario, string) {
+	return &wan{}, GetScenarioName()
 }

-func New(name string) scenarios.Scenario {
-	return &wan{
-		name: "mesh",
-	}
+func GetScenarioName() string {
+	return "WAN_FEDERATION"
 }

 func (w *wan) GetFolderName() string {
 	return "mesh-gateways"
 }

+func (w *wan) GetTerraformResourcesName() string {
+	return "mesh"
+}
+
 func (w *wan) GetTerraformVars() (map[string]interface{}, error) {
 	vars := map[string]interface{}{
 		"region": "us-east-1",
-		"name":   w.name,
+		"name":   w.GetTerraformResourcesName(),
 	}

 	publicIP, err := common.GetPublicIP()
@@ -65,13 +69,13 @@ func (w *wan) Validate(t *testing.T, outputVars map[string]interface{}) {
 	consulClientTwo, err := common.SetupConsulClient(t, dc2ConsulServerURL, common.WithToken(dc1ConsulServerToken))
 	require.NoError(t, err)

-	clientAppName := fmt.Sprintf("%s-dc1-example-client-app", w.name)
-	serverAppName := fmt.Sprintf("%s-dc2-example-server-app", w.name)
+	clientAppName := fmt.Sprintf("%s-dc1-example-client-app", w.GetTerraformResourcesName())
+	serverAppName := fmt.Sprintf("%s-dc2-example-server-app", w.GetTerraformResourcesName())

 	consulClientOne.EnsureServiceReadiness(clientAppName, nil)
 	consulClientTwo.EnsureServiceReadiness(serverAppName, nil)
-	consulClientOne.EnsureServiceReadiness(fmt.Sprintf("%s-dc1-mesh-gateway", w.name), nil)
-	consulClientTwo.EnsureServiceReadiness(fmt.Sprintf("%s-dc2-mesh-gateway", w.name), nil)
+	consulClientOne.EnsureServiceReadiness(fmt.Sprintf("%s-dc1-mesh-gateway", w.GetTerraformResourcesName()), nil)
+	consulClientTwo.EnsureServiceReadiness(fmt.Sprintf("%s-dc2-mesh-gateway", w.GetTerraformResourcesName()), nil)

 	// Perform assertions by hitting the client app's LB
 	logger.Log(t, "calling client app's load balancer to see if the server app in the secondary DC is reachable")
```

Reply:
Update - 
I have added two methods to Scenario Interface - `GetScenarioName` and `GetTerraformResourcesName`

Reply:
I wanted to get rid of this hardcoding as well and wanted to delegate as much as possible to the individual scenarios. I have decided to move away from the interface pattern for registering the scenario and added a registry to keep track of the individual scenarios. Now the `TestScenario` function just needs to interact with the registry and operate on the scenario. https://github.com/hashicorp/terraform-aws-consul-ecs/pull/250/commits/779963ad4bd409f41788492bc8bc10c7163e43a3 should give you an idea about the new pattern. I have only modified this for FARGATE but will make changes for others too in the upcoming commits.

Reply:
This looks good, similar to V2. 😁